gslib/gcs_json_api.py

```diff
@@ -1162,14 +1194,20 @@ def GetObjectMedia(self,
     self.download_http.connections = {'https': download_http_class}
 
     if serialization_data:
+      serialization_dict = json.loads(six.ensure_str(serialization_data))
+
       # If we have an apiary trace token, add it to the URL.
       # TODO: Add query parameter support to apitools downloads so there is
       # a well-defined way to express query parameters. Currently, we assume
       # the URL ends in ?alt=media, and this will break if that changes.
       if self.trace_token:
-        serialization_dict = json.loads(six.ensure_str(serialization_data))
         serialization_dict['url'] += '&trace=token%%3A%s' % self.trace_token
-        serialization_data = json.dumps(serialization_dict)
+
+      # Maintain custom endpoints that might have been overrided.
+      serialization_dict['url'] = self.api_client.FinalizeTransferUrl(
```

Comment:
Do you need to call this here? Won't this be called by Apitools anyway given that you have patched the method already on line 338?

Reply:
Part of the cause of this bug is that Apitools doesn't call `FinalizeTransferUrl` on URLs received through `FromData`.

Reply:
So in that case, should there be a fix in Apitools?

Reply:
It might be an intentional Apitools decision, and it could break other things, so I'd say no